Implement write() method for Box<MaybeUninit<T>>

Kixunil opened the pull request:
This adds method similar to `MaybeUninit::write` main difference being
it returns owned `Box`. This can be used to elide copy from stack
safely, however it's not currently tested that the optimization actually
occurs.

Analogous methods are not provided for `Rc` and `Arc` as those need to
handle the possibility of sharing. Some version of them may be added in
the future.

This was discussed in #63291 which this change extends.

rust-highfive commented:
r? @dtolnay

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3587784977) [(plain)](https://github.com/rust-lang/rust/commit/005cd4bf8908f90ed0943cd515294c28fb91c687/checks/3587784977/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking alloc v0.0.0 (/checkout/library/alloc)
    Checking cfg-if v0.1.10
    Checking adler v0.2.3
    Checking rustc-demangle v0.1.21
error[E0596]: cannot borrow `*self` as mutable, as `self` is not declared as mutable
    |
781 |     pub fn write(self, value: T) -> Box<T, A> {
781 |     pub fn write(self, value: T) -> Box<T, A> {
    |                  ---- help: consider changing this to be mutable: `mut self`
782 |         unsafe {
783 |             (*self).write(value);
    |             ^^^^^^^ cannot borrow as mutable
For more information about this error, try `rustc --explain E0596`.
error: could not compile `alloc` due to previous error
Build completed unsuccessfully in 0:01:21
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3587852647) [(plain)](https://github.com/rust-lang/rust/commit/d0c2ccf204df00afb8cd69aa1d6aaa7776f1c8ae/checks/3587852647/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc-demangle v0.1.21
error: unknown start of token: `
 --> library/alloc/src/boxed.rs:3:6
  |
3 | 5/! [`Box<T>`], casually referred to as a 'box', provides the simplest form of
  |
  |
help: Unicode character '`' (Grave Accent) looks like ''' (Single Quote), but it is not
  |
3 | 5/! ['Box<T>`], casually referred to as a 'box', provides the simplest form of

error: unknown start of token: `
 --> library/alloc/src/boxed.rs:3:13
  |
  |
3 | 5/! [`Box<T>`], casually referred to as a 'box', provides the simplest form of
  |
  |
help: Unicode character '`' (Grave Accent) looks like ''' (Single Quote), but it is not
  |
3 | 5/! [`Box<T>'], casually referred to as a 'box', provides the simplest form of


error: character literal may only contain one codepoint
  |
  |
3 | 5/! [`Box<T>`], casually referred to as a 'box', provides the simplest form of
  |
  |
help: if you meant to write a `str` literal, use double quotes
  |
3 | 5/! [`Box<T>`], casually referred to as a "box", provides the simplest form of

error: expected item, found `5`
 --> library/alloc/src/boxed.rs:3:1
  |
  |
3 | 5/! [`Box<T>`], casually referred to as a 'box', provides the simplest form of

error[E0432]: unresolved import `crate::boxed::Box`
  --> library/alloc/src/collections/btree/node.rs:40:5
   |
   |
40 | use crate::boxed::Box;
   |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
  --> library/alloc/src/collections/linked_list.rs:24:5
   |
24 | use crate::boxed::Box;
24 | use crate::boxed::Box;
   |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
  --> library/alloc/src/prelude/v1.rs:10:9
   |
10 | pub use crate::boxed::Box;
10 | pub use crate::boxed::Box;
   |         ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
  --> library/alloc/src/raw_vec.rs:15:5
   |
15 | use crate::boxed::Box;
15 | use crate::boxed::Box;
   |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
   --> library/alloc/src/rc.rs:246:5
    |
246 | use crate::boxed::Box;
246 | use crate::boxed::Box;
    |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
   --> library/alloc/src/slice.rs:100:5
    |
100 | use crate::boxed::Box;
100 | use crate::boxed::Box;
    |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
   --> library/alloc/src/slice.rs:155:9
    |
155 |     use crate::boxed::Box;
155 |     use crate::boxed::Box;
    |         ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
  --> library/alloc/src/str.rs:38:5
   |
38 | use crate::boxed::Box;
38 | use crate::boxed::Box;
   |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
  --> library/alloc/src/string.rs:67:5
   |
67 | use crate::boxed::Box;
67 | use crate::boxed::Box;
   |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
  --> library/alloc/src/sync.rs:36:5
   |
36 | use crate::boxed::Box;
36 | use crate::boxed::Box;
   |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
  --> library/alloc/src/vec/mod.rs:74:5
   |
74 | use crate::boxed::Box;
74 | use crate::boxed::Box;
   |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error[E0432]: unresolved import `crate::boxed::Box`
 --> library/alloc/src/vec/is_zero.rs:1:5
  |
1 | use crate::boxed::Box;
1 | use crate::boxed::Box;
  |     ^^^^^^^^^^^^^^^^^ no `Box` in `boxed`
error: specializing impl repeats parameter `T`
    --> library/alloc/src/rc.rs:1896:1
     |
     |
1896 | / impl<T, const N: usize> TryFrom<Rc<[T]>> for Rc<[T; N]> {
1897 | |     type Error = Rc<[T]>;
1898 | |
1899 | |     fn try_from(boxed_slice: Rc<[T]>) -> Result<Self, Self::Error> {
1905 | |     }
1906 | | }
     | |_^


error: specializing impl repeats parameter `T`
    --> library/alloc/src/sync.rs:2486:1
     |
2486 | / impl<T, const N: usize> TryFrom<Arc<[T]>> for Arc<[T; N]> {
2487 | |     type Error = Arc<[T]>;
2488 | |
2489 | |     fn try_from(boxed_slice: Arc<[T]>) -> Result<Self, Self::Error> {
2495 | |     }
2496 | | }
     | |_^

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3587957093) [(plain)](https://github.com/rust-lang/rust/commit/322b67dc54958d32333e7fd385e96c96d824ea67/checks/3587957093/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_data_structures v0.0.0 (/checkout/compiler/rustc_data_structures)
error: variable does not need to be mutable
  --> compiler/rustc_data_structures/src/functor.rs:27:17
   |
27 |             let mut raw: Box<mem::MaybeUninit<T>> = Box::from_raw(raw.cast());
   |                 |
   |                 help: remove this `mut`
   |
   |
   = note: `-D unused-mut` implied by `-D warnings`
error[E0382]: use of moved value: `raw`
  --> compiler/rustc_data_structures/src/functor.rs:31:13
   |
   |
27 |             let mut raw: Box<mem::MaybeUninit<T>> = Box::from_raw(raw.cast());
   |                 ------- move occurs because `raw` has type `Box<MaybeUninit<T>>`, which does not implement the `Copy` trait
28 |             // SAFETY: Write the mapped value back into the `Box`.
29 |             raw.write(f(value));
   |             --- value moved here
30 |             // SAFETY: We just initialized `raw`.
31 |             raw.assume_init()
   |             ^^^ value used here after move
For more information about this error, try `rustc --explain E0382`.
error: could not compile `rustc_data_structures` due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

the8472 commented:
There's a pending RFC that proposes another approach which would probably make this one redundant.

https://github.com/rust-lang/rfcs/pull/2884

Kixunil commented:
@the8472 thanks for heads up! It could be argued it also makes #6321 redundant (although that one can be useful in FFI but then why not add this one for completeness?)

dtolnay reviewed (commented):
> I believe this change is simple and obvious.

To point out why this change isn't simple and obvious: this is a breaking change. It can make previously working code fail to compile, or keep compiling but do the wrong thing, or silently introduce UB into existing correct code.

Here is an example of an ordinary idiomatic use of MaybeUninit which would no longer work:

```rust
struct Thing {
    buf: Box<MaybeUninit<[u8; N]>>,
}

impl Thing {
    fn method(&mut self) {
        let _arr = self.buf.write([0; N]);
    }
}
```

Kixunil commented:
Ah, I see, good point. I think changing it to `fn write(this: Self)` would fix it, right? Will try to do it when I have time (shouldn't be more than couple of days).

bors commented:
:umbrella: The latest upstream changes (presumably #91230) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Kixunil commented:
Fixed and rebased. Also noticed that wording in the example looked like a promise of optimization which is not actually guaranteed, so I reworded it accordingly.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4397630088?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/928a0b00ae919ffe667f9f90ade82ef572cbe89a/checks/4397630088/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  --> compiler/rustc_data_structures/src/functor.rs:29:11
   |
27 |     {
   |     - closing delimiter possibly meant for this
28 |         let raw = Box::into_raw(self);
29 |         Ok(unsafe {
...
38 |     }
   |     ^ mismatched closing delimiter

```

</details>

dtolnay reviewed (approved):
Thanks!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 41e21aa1c27522c6b29c5b88e37b1f479f63e38c has been approved by `dtolnay`

<!-- @bors r=dtolnay 41e21aa1c27522c6b29c5b88e37b1f479f63e38c -->
<!-- homu: {"type":"Approved","sha":"41e21aa1c27522c6b29c5b88e37b1f479f63e38c","approver":"dtolnay"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #91486) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Kixunil commented:
I'm confused by this - it conflicted with an unmerged PR? Is this some fluke or am I actually supposed to do something (what?) about it?

pietroalbini commented:
This has been merged into master by bors (in https://github.com/rust-lang/rust/pull/91486), but GitHub didn't detect it for whatever reason. Closing.
